core/src/test/resources/tck-test-files/cypher/advanced-filtering.adoc

```diff
@@ -0,0 +1,649 @@
+----
+
+=== LTE
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { actorCount_lte: 123 }) {
+        actorCount
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+  "whereMovieActorCountLte" : 123
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (movie:Movie)
+WHERE movie.actorCount <= $whereMovieActorCountLte
+RETURN movie {
+	.actorCount
+} AS movie
+----
+
+=== LTE BigInt
+
+CAUTION: *Not yet implemented*
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { budget_lte: 9223372036854775807 }) {
+        budget
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+    "this_budget_LTE": {
+        "low": -1,
+        "high": 2147483647
+    }
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (this:Movie)
+WHERE this.budget <= $this_budget_LTE
+RETURN this { .budget } as this
+----
+
+=== GT
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { actorCount_gt: 123 }) {
+        actorCount
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+  "whereMovieActorCountGt" : 123
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (movie:Movie)
+WHERE movie.actorCount > $whereMovieActorCountGt
+RETURN movie {
+	.actorCount
+} AS movie
+----
+
+=== GT BigInt
+
+CAUTION: *Not yet implemented*
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { budget_gt: 9223372036854775000 }) {
+        budget
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+    "this_budget_GT": {
+        "low": -808,
+        "high": 2147483647
+    }
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (this:Movie)
+WHERE this.budget > $this_budget_GT
+RETURN this { .budget } as this
+----
+
+=== GTE
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { actorCount_gte: 123 }) {
+        actorCount
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+  "whereMovieActorCountGte" : 123
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (movie:Movie)
+WHERE movie.actorCount >= $whereMovieActorCountGte
+RETURN movie {
+	.actorCount
+} AS movie
+----
+
+=== GTE BigInt
+
+CAUTION: *Not yet implemented*
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { budget_gte: 9223372036854775000 }) {
+        budget
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+    "this_budget_GTE": {
+        "low": -808,
+        "high": 2147483647
+    }
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (this:Movie)
+WHERE this.budget >= $this_budget_GTE
+RETURN this { .budget } as this
+----
+
+=== Relationship equality
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+    movie(where: { genres: { name: "some genre" } }) {
+        actorCount
+    }
+}
+----
+
+.Expected Cypher params
+[source,json]
+----
+{
+  "whereMovieGenreName" : "some genre"
+}
+----
+
+.Expected Cypher output
+[source,cypher]
+----
+MATCH (this:Movie)
```

Comment:
@jexp This is what the JS library is currently generating.

Our code would generate:

```cypher
MATCH (movie:Movie)
WHERE all(whereMovieGenreCond IN [(movie)-[:IN_GENRE]->(whereMovieGenre:Genre) | whereMovieGenre.name = $whereMovieGenreName]
WHERE whereMovieGenreCond)
RETURN movie {
	.actorCount
} AS movie
```

Should I migrate to the js version or should I keep our logic?

Reply:
Ours is better as the planner is able to pull the condition into the expansion and doesn't have to build up a full list and post-filter, perhaps we can raise an issue with their repo to point that out.

Also we should really investigate the subqueries approach both with existential subqueries but also real subqueries to see the (hopefully positive) impact for the planner.

Reply:
done https://github.com/neo4j/graphql/issues/187